tests/Microsoft.Bot.Connector.Tests/Authentication/JwtTokenExtractorTests.cs

```diff
@@ -0,0 +1,156 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.IdentityModel.Tokens.Jwt;
+using System.Net.Http;
+using System.Security.Claims;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Bot.Connector.Authentication;
+using Microsoft.Bot.Schema;
+using Microsoft.IdentityModel.Tokens;
+using Xunit;
+
+namespace Microsoft.Bot.Connector.Tests.Authentication
+{
+    public class JwtTokenExtractorTests
+    {
+        private readonly HttpClient client;
+        private readonly HttpClient emptyClient;
+
+        public JwtTokenExtractorTests()
+        {
+            // Disable TokenLifetime validation
+            EmulatorValidation.ToBotFromEmulatorTokenValidationParameters.ValidateLifetime = false;
+            ChannelValidation.ToBotFromChannelTokenValidationParameters.ValidateLifetime = false;
+            GovernmentChannelValidation.ToBotFromGovernmentChannelTokenValidationParameters.ValidateLifetime = false;
+            client = new HttpClient
+            {
+                BaseAddress = new Uri("https://webchat.botframework.com/"),
+            };
+            emptyClient = new HttpClient();
+        }
+
+        [Fact]
+        public void JwtTokenExtractor_GovernmentValidationParameters_ShouldValidateSigningKey()
+        {
+            Assert.True(GovernmentChannelValidation.ToBotFromGovernmentChannelTokenValidationParameters.ValidateIssuerSigningKey);
+        }
+
+        [Fact]
+        public void JwtTokenExtractor_ChannelValidationParameters_ShouldValidateSigningKey()
+        {
+            Assert.True(ChannelValidation.ToBotFromChannelTokenValidationParameters.ValidateIssuerSigningKey);
+        }
+
+        [Fact]
+        public void JwtTokenExtractor_EnterpriseChannelValidationParameters_ShouldValidateSigningKey()
+        {
+            Assert.True(EnterpriseChannelValidation.ToBotFromEnterpriseChannelTokenValidationParameters.ValidateIssuerSigningKey);
+        }
+
+        [Fact]
+        [Trait("Platform", "WindowsOnly")]
+        public async Task JwtTokenExtractor_WithExpiredCert_ShouldNotAllowCertSigningKey()
+        {
+            var now = DateTimeOffset.UtcNow;
+
+            // Create expired self-signed certificate
+            var cert = CreateSelfSignedCertificate("test.cert.botframework.com", from: now.AddDays(-10), to: now.AddDays(-9));
```

Comment:
FYI this will create a file on disk with the private key container in %APPDATA%\Microsoft\Crypto\RSA - I think you can clean up with this code:

            public void DeleteKeyContainer()
            {
                // %APPDATA%\Microsoft\Crypto\RSA
                var parameters = new CspParameters()
                {
                    KeyContainerName = Name,
                    Flags = CspProviderFlags.NoPrompt | CspProviderFlags.UseExistingKey,
                };

                try
                {
                    using (var provider = new RSACryptoServiceProvider(parameters))
                    {
                        provider.PersistKeyInCsp = false;
                        provider.Clear();
                    }
                }
                catch (CryptographicException error) when (CertificateScope.IsKeySetDoesNotExist(error))
                {
                }
            }

        private const int NTE_BAD_KEYSET = unchecked((int)0x80090016);

        public static bool IsKeySetDoesNotExist(Exception error) =>
            error.HResult == NTE_BAD_KEYSET;

Reply:
Thanks! Adding the cleanup now.

Reply:
Added, thanks for the code :)